### ISSUE TITLE
Extend elastic_vector_search.py to allow for kNN indexing/searching

### ISSUE THREAD

**Issue by jeffvestal** [hwchase17/langchain#5346]
### Feature request

Extend `langchain/embeddings/elasticsearch.py` to support kNN indexing and searching. 
The high-level objectives will be:

1. Allow for the [creation of an index with the correct mapping](https://www.elastic.co/guide/en/elasticsearch/reference/current/dense-vector.html#index-vectors-knn-search) to store documents including dense_vectors so they can be used for kNN search
2. Store embeddings in elasticsearch in [dense_vector](https://www.elastic.co/guide/en/elasticsearch/reference/current/dense-vector.html) field type
3. Perform [kNN search](https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html)
4. Perform [Hybrid](https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html#_combine_approximate_knn_with_other_features) BM25 (query) + kNN search

### Motivation

Elasticsearch support approximate k-nearest neighbor search with dense vectors. The current module only support script score / exact match vector search. 

### Your contribution

I will work on the code and create the pull request

**Comment by ManZzup** [comment]
Had this need today and did a crud version of it, let me know if this make sense - I can open a PR with it
I also added batching for bulk add - it was failing for larger datasets

https://gist.github.com/ManZzup/1109d4c1f6b8bc48b60a67983dfbd0fd

**Comment by jeffvestal** [comment]
I'm almost done with adding a new `class ElasticKnnSearch`
[Working code is here](https://github.com/jeffvestal/langchain/blob/62548ce16ae0787f400fa69c3192636ce8cd090c/langchain/vectorstores/elastic_vector_search.py#L286) if you want to take a look 

I've been busy on a couple different projects but I'm hoping to have the pr sumbitted in the next day or two. Just debugging and cleaning up right now

It you do take a look feel free to comment here if you have any suggestions

**Comment by ManZzup** [comment]
@jeffvestal this looks much better and structured! Will wait for this to merge to use this instead.

A couple of things,
- Would it be good to let user define the boost? For my use case for instance 0.5 and 0.5 made more sense
- I added a source filter to prevent the `vector` field from coming back in response. Since we are only using the `text` and `metadata` fields from the response, do we still need to retain the vector? (I saw you have `fields` commented, i think newer API supports the `source` field instead)
-
